x/incentive/keeper/refund_tx_decorator.go

```diff
@@ -0,0 +1,64 @@
+package keeper
+
+import (
+	"github.com/babylonlabs-io/babylon/x/incentive/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+var _ sdk.PostDecorator = &RefundTxDecorator{}
+
+type RefundTxDecorator struct {
+	k *Keeper
+}
+
+// NewRefundTxDecorator creates a new RefundTxDecorator
+func NewRefundTxDecorator(k *Keeper) *RefundTxDecorator {
+	return &RefundTxDecorator{
+		k: k,
+	}
+}
+
+func (d *RefundTxDecorator) PostHandle(ctx sdk.Context, tx sdk.Tx, simulate, success bool, next sdk.PostHandler) (sdk.Context, error) {
+	// refund only when finalizing a block or simulating the current tx
+	if ctx.ExecMode() != sdk.ExecModeFinalize && !simulate {
+		return next(ctx, tx, simulate, success)
+	}
+	// avoid refunding failed tx
+	if !success {
+		return next(ctx, tx, simulate, success)
+	}
+	// ignore tx that does not need to pay fee
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return next(ctx, tx, simulate, success)
+	}
+
+	refundableMsgHashList := make([][]byte, 0)
+
+	// iterate over all messages in the tx, and record whether they are refundable
+	for _, msg := range tx.GetMsgs() {
+		msgHash := types.HashMsg(msg)
+		if d.k.HasRefundableMsg(ctx, msgHash) {
+			refundableMsgHashList = append(refundableMsgHashList, msgHash)
+		}
+	}
+
+	// if all messages in the tx are refundable, refund the tx
+	if len(refundableMsgHashList) == len(tx.GetMsgs()) {
+		err := d.k.RefundTx(ctx, feeTx)
+		if err != nil {
```

Comment:
why not return the error if it fails to refund?

Reply:
If this returns error then the tx will be rejected and all resulting state transition will be reverted. It's a bit unexpected to fail the tx due to fee refund, but tbh not sure if it's a good thing to do. If `RefundTx` fails it might mean something is really wrong. Will think more about it tmr